README
-------------------------------------------------------------------

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                            TestEm3
                            -------
  
 How to collect energy deposition in a sampling calorimeter.
 How to survey energy flow.
 how to print stopping power. 

 1- GEOMETRY DEFINITION
 
  The calorimeter is a box made of a given number of layers.
  A layer consists of a sequence of various absorbers (maximum MaxAbsor=9).
  The layer is replicated.
 
  Parameters defining the calorimeter :
    - the number of layers,
    - the number of absorbers within a layer,		
    - the material of the absorbers,
    - the thickness of the absorbers,
    - the transverse size of the calorimeter (the input face is a square). 
 
  In addition a transverse uniform magnetic field can be applied.
 
  The default geometry is constructed in DetectorConstruction class, but all
  of the above parameters can be modified interactively via the commands 
  defined in the DetectorMessenger class.


        |<----layer 0---------->|<----layer 1---------->|<----layer 2---------->|
        |           |           |                       |                       |
        ==========================================================================
        ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
        ||   abs 1  | abs 2     ||   abs 1  | abs 2     ||   abs 1  | abs 2     ||
        ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
 beam   ||          |           ||          |           ||          |           ||
======> ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
        ||          |           ||          |           ||          |           ||
        ||   cell 1 | cell 2    ||   cell 3 | cell 4    ||   cell 5 | cell 6    ||
        ==========================================================================
        ^           ^           ^           ^           ^           ^           ^
        pln1        pln2        pln3       pln4        pln5        pln6       pln7
 
  NB. The number of absorbers and the number of layers can be set to 1.
  In this case we have a unique homogeneous block of matter, which looks like 
  a bubble chamber rather than a calorimeter ...
  (see the macro emtutor.mac)
 
 2- PHYSICS LISTS
 
  Physics lists are based on modular design. Several modules are instantiated:
  1. Transportation
  2. EM physics
  3. Decays
  4. StepMax - for step limitation

  EM physics builders can be local (eg. in this example) or from G4 kernel
  physics_lists subdirectory.

  Local physics builders:	 
  - "local"	standard EM physics with current 'best' options setting.
              these options are explicited in PhysListEmStandard
  
  From geant4/source/physics_lists/builders:	
  - "emstandard_opt0" recommended standard EM physics for LHC
  - "emstandard_opt1" best CPU performance standard physics for LHC
  - "emstandard_opt2" similar fast simulation
  - "emstandard_opt3" best standard EM options - analog to "local" above
  - "emstandard_opt4" best current advanced EM options standard + lowenergy
  - "emstandardWVI" standard EM physics and WentzelVI multiple scattering
  - "emstandardSS"  standard EM physics and single scattering model
  - "emlivermore"  low-energy EM physics using Livermore data
  - "empenelope"   low-energy EM physics implementing Penelope models
  - "emlowenergy"  low-energy EM physics implementing experimental
                    low-energy models
      
  Physics lists and options can be (re)set with UI commands.
    
 3- AN EVENT : THE PRIMARY GENERATOR
 
  The primary kinematic consists of a single particle which hits the calorimeter
  perpendicular to the input face. The type of the particle and its energy are 
  set in the PrimaryGeneratorAction class, and can be changed via the 
  G4 build-in commands of G4ParticleGun class (see the macros provided with this 
  example).
 	
  In addition one can choose randomly the impact point of the incident particle.
  The corresponding interactive command is built in PrimaryGeneratorMessenger.
 	
  A RUN is a set of events.
  
  TestEm3 computes the energy deposited per absorber and the energy flow through
  the calorimeter.
 				
 4- VISUALIZATION
 
  The Visualization Manager is set in the main() (see TestEm3.cc).
  The initialisation of the drawing is done via the commands :
  /vis/... in the macro vis.mac. In interactive session:
  PreInit or Idle > /control/execute vis.mac
 	
  The default view is a longitudinal view of the calorimeter.
 	
  The tracks are drawn at the end of event, and erased at the end of run.
  Optionally one can choose to draw all particles, only the charged ones, or
  none. This command is defined in EventActionMessenger class.
 	
 5- PHYSICS DEMO
 
  The particle's type and the physics processes which will be available
  in this example are set in PhysicsList class.
 	
  In addition a built-in interactive command (/process/inactivate processName)
  allows to activate/inactivate the processes one by one.
  Then one can well visualize the processes one by one, especially 
  in the bubble chamber setup with a transverse magnetic field.
 
  As a homework try to visualize a gamma conversion alone, 
  or the effect of the multiple scattering.

  Notice that one can control the maximum step size, via the
  StepMax process and the command /testem/stepMax
  (see StepMax and PhysicsList classes)
 
 6- HOW TO START ?
 
  - Execute TestEm3 in 'batch' mode from macro files
      % TestEm3   run01.mac
 
  - Execute TestEm3 in 'interactive mode' with visualization
    % TestEm3
    ....
    Idle> type your commands. For instance:
    Idle> /control/execute run01.mac
    ....
    Idle> exit

  Macros provided in this example:
  - atlashec.mac: ATLAS HEC model
  - dedx.mac: to control dE/dx calculation: 1 layer; minimum ionizing particle
  - emtutor.mac: for tutorial; interactivity + visualisation
  - ionC12.mac: ion C12, 1 layer
  - lhcb.mac: LHCB ECAL model
  - linac.mac: Linac/Ecal from Graham Wilson
  - lockwood.mac: Al-Au-Al 1 layer (G.L.Lockwood et al. SAND79-0414 (1980))
  - run01.mac: Lead-liquidArgon 50 layers; electron 1 GeV
  - run02.mac: Tungsten-Silicon 50 layers; electron 1 GeV
  - storeTables.mac: show how to store and retrieve physics tables
  - tileCal.mac: ATLAS tileCal
  - vis.mac: to activate visualization

 7- HISTOGRAMS
 
 Testem3 can produce histograms : 
  histo 1 : energy deposit in absorber 1
  histo 2 : energy deposit in absorber 2
  ...etc...........
    
  histo 11 : longitudinal profile of energy deposit in absorber 1 (MeV/event)
  histo 12 : longitudinal profile of energy deposit in absorber 2 (MeV/event)  
  ...etc...........  
  
  histo 21 : energy flow (MeV/event)
  histo 22 : lateral energy leak (MeV/event)  
  ...etc...........
      
  NB. Numbering scheme for histograms:
  layer     : from 1 to NbOfLayers (included)
  absorbers : from 1 to NbOfAbsor (included)
  planes    : from 1 to NbOfLayers*NbOfAbsor + 1 (included)
  
 One can control the binning of the histo with the command:
  /analysis/h1/set   idAbsor  nbin  Emin  Emax  unit 
  where unit is the desired energy unit for that histo (see TestEm3.in).
         
  One can control the name of the histograms file with the command:
  /analysis/setFileName  name  (default testem3)
   
  It is possible to choose the format of the histogram file : root (default),
  xml, csv, by using namespace in HistoManager.hh 
   	
 It is also possible to print selected histograms on an ascii file:
 /analysis/h1/setAscii id
 All selected histos will be written on a file name.ascii  (default testem3)
    


History
# Example TestEm3 History

See `CONTRIBUTING.rst` for details of **required** info/format for each entry,
which **must** added in reverse chronological order (newest at the top). It must **not**
be used as a substitute for writing good git commit messages!

## 2023-01-23 Michel Maire (testem3-V11-01-00)
- compute and plot energy leakage and total energy

## 2022-01-18 Jonas Hahnfeld (testem3-V11-00-01)
- Prefer pointer to `const G4Material` if possible

## 2021-12-10 Ben Morgan (testem3-V11-00-00)
- Change to new Markdown History format

---

# History entries prior to 11.0
     
06-10-21 I. Hrivnacova (testem3-V10-07-03)
- Migration to new G4AnalysisManager.hh header;
  define the default output file type (root).

19-07-21 I. Hrivnacova (testem3-V10-07-02)
- Updated for changes in the analysis category:
  removed deleting of the analysis manager,
  as this is now done by the Geant4 kernel.

06-05-21 mma (testem3-V10-07-01)
- Migration to G4RunManagerFactory and G4SteppingVerboseWithUnits.
       
14-12-20 mma (testem3-V10-07-00)
- updated PhysListEmStandard.cc
- StepMax, StepMaxMessenger, PhysicsList, PhysicsListMessenger :
    revert, to make command /testem/stepMax working in interactive mode
- added macro stepMax.mac
      
14-09-20 I. Hrivnacova (testem3-V10-06-01)
- Renamed TestEm3_2.in in geom.mac

01-09-20 mma, I. Hrivnacova (testem3-V10-06-00)
- Macro & commands review:
  - Added macro descriptions in README
  - Added TestEm3_2.in with commands not used in TestEm3.in
  - Fixed not working macros: ionC12.mac, lockwood.mac, storeTables.mac
  - Removed obsolete macro retrieveTables.mac 

26-11-19 I. Hrivnacova (testem3-V10-05-00)
- Fixed Doxygen warnings

09-05-18 B.Morgan (testem3-V10-04-06)
- Include G4Types before use of G4MULTITHREADED. For forward
  compatibility with move to #defines over -D for G4 preprocessor
  symbols.

29-03-18 V.Ivant (testem3-V10-04-05)
- TestEm3.cc - small optimisation

27-03-18 mma (testem3-V10-04-04)
- remove radioactive.mac

26-03-18 V.Ivant (testem3-V10-04-03)
- DetectorConstruction - fixed initialisation

23-03-18 V.Ivant (testem3-V10-04-02)
- StepMax - use thread safe instantiation, mesenger merged with 
    PhysicsListMessenger
- DetectorConstruction - improved method names, not allow to clear 
    geometry between runs
- PhysicsList - make Opt0 default, removed local instantiation of G4Decay
- Use MixMax default random number generator 

13-03-18 mma (testem3-V10-04-01)
- testem1.cc : remove G4UI_USE and G4VIS_USE

14-12-17 V.Ivant (testem3-V10-04-00)
- corrected TestEm3.in to optimise output

14-10-16 G.Folger (testem3-V10-02-06)
- remove direct use of theParticleIterator, use GetParticleTableIterator().
    fix required by clang39 on Linux and MAC

22-07-16 mma  (testem3-V10-02-05)
- apply coding conventions

04-07-16 I. Hrivnacova  (testem3-V10-02-04)
- Fixes for Doxygen documentation

10-02-16 mma (testem3-V10-02-03)
- update radioactive.mac again

06-02-16 mma (testem3-V10-02-02)
- update radioactive.mac

29-01-16 mma (testem3-V10-02-01)
- Physicsist: add radioactiveDecay process (not active by default)

10-12-15 V.Ivant (testem3-V10-02-00)
- H.Burkhardt fixed root macro allowing to work both with root5 and 6
    
28-10-15 D.Sawkey (testem3-V10-01-04)
- update description of physics lists in README, again

26-10-15 D.Sawkey (testem3-V10-01-03)
- update description of physics lists in README

19-09-15 V.Ivant (testem3-V10-01-02)
- TestEm3.cc - use default 4 threads, improve vis.mac
- PhysListEmStandard - use G4EmParameters

12-05-15 V.Ivant (testem3-V10-01-01)
- PhysicsList - use G4EmStandardPhysicsGS from physics_list library

23-01-15 I. Hrivnacova (testem3-V10-01-00)
- fixed formatting in .README
	 
28-10-14 mma (testem3-V10-00-16)
- in macros : /run/numberOfThreads 2
 
23-10-14 V.Ivant (testem3-V10-00-15)
- disable extra printout for sequential 
	 
23-10-14 V.Ivant (testem3-V10-00-14)
- improved output MT 
	 
23-10-14 mma (testem3-V10-00-13)
- PhysListEmStandard : msc algorithm fUseSafetyPlus

21-10-14 mma (testem3-V10-00-12)
- PhysicsList : added G4EmStandardPhysicsSS
- TestEm3.cc : option set nThreads via command
	 
27-08-14 mma (testem3-V10-00-11)
- TestEm1.cc : SetNumberOfThreads( G4Threading::G4GetNumberOfCores() )

13-08-14 mma (testem3-V10-00-10)
- update .README again

12-08-14 mma (testem3-V10-00-09)
- update .README

11-08-14 V.Ivant (testem3-V10-00-08)
- added "emstandardSS" option for EM physics 
	 
10-08-14 mma (testem3-V10-00-07)
- update README, emtutor.mac
		 
14-06-14 mma (testem3-V10-00-06)
- remove PhysicsList::SetCuts()	
 
11-06-14 mma (testem3-V10-00-05)
- retag of 04
 
07-06-14 mma (testem3-V10-00-04)
- minor cleanup in PhysicsList::SetCuts()
	 
05-06-14 mma (testem3-V10-00-03)
- replace /testem/phys/setCuts by /run/setCut
 
28-04-14 V.Ivant (testem3-V10-00-02)
- added "emlowenergy" option for EM physics 
     
14-01-14 mma (testem3-V10-00-01)
- MT: Add forgotten classes ActionInitialization, Run
     
06-01-14 mma (testem3-V10-00-00)
- MT: ActionInitialization, TestEm3.cc
- MT: Run, RunAction, EventAction, TrackingAction, SteppingAction
- MT: DetectorConstruction and Messenger
- rm EventActionMessenger, hbook

24-07-13 P.Gumplinger (testem3-V09-06-09)
- remove unused field fEventAct from TrackingAction class

11-07-13 I.Hrivnacova (testem3-V09-06-08) 
- Updated CMakeLists.txt and SharedFilesList.txt for added classes
  in Hbook analysis code
- In HistoManager: use new analysis manager method SetH1Activation()   
    
19-06-13 mma (testem3-V09-06-07)
- SteppingVerbose: use G4Step::GetSecondaryInCurrentStep() .
     
06-06-13 mma (testem3-V09-06-06)
- Cosmetic in SteppingAction.

14-04-13 I.Hrivnacova (testem3-V09-06-05)
- Removed ExG4HbookAnalysisManager and related files and updated CMake/GNUmake
  files to use it from extended/common          
     
01-04-13 A.Dotti (testem3-V09-06-02,testem3-V09-06-03,testem3-V09-06-04)
- New G4VUserPhysicsList interface 
     
20-03-13 mma (testem3-V09-06-01)

src/HistoManager.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
//
// 
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "HistoManager.hh"
#include "G4UnitsTable.hh"
#include "DetectorConstruction.hh"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

HistoManager::HistoManager()
  : fFileName("testem3")
{
  Book();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

HistoManager::~HistoManager()
{
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void HistoManager::Book()
{
  // Create or get analysis manager
  // The choice of analysis technology is done via selection of a namespace
  // in HistoManager.hh
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
  analysisManager->SetDefaultFileType("root");
  analysisManager->SetFileName(fFileName);
  analysisManager->SetVerboseLevel(1);
  analysisManager->SetActivation(true);   // enable inactivation of histograms

  // Define histograms start values
  
  const G4String id[] = { "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
                         "10","11","12","13","14","15","16","17","18","19",
                         "20","21","22","23","24"};
  G4String title;

  // Default values (to be reset via /analysis/h1/set command)
  G4int nbins = 100;
  G4double vmin = 0.;
  G4double vmax = 100.;
  
  // Create all histograms as inactivated 
  // as we have not yet set nbins, vmin, vmax
  for (G4int k=0; k<kMaxHisto; k++) {
    if (k < kMaxAbsor)  title = "Edep in absorber " + id[k];
    if (k == kMaxAbsor) title = "total energy deposited";
    if (k > kMaxAbsor)  title = "Edep longit. profile (MeV/event) in absorber "
                               + id[k-kMaxAbsor];
    if (k == 2*kMaxAbsor+1) title = "energy flow (MeV/event)";
    if (k == 2*kMaxAbsor+2) title = "lateral energy leak (MeV/event)";
    if (k == 2*kMaxAbsor+3) title = "total energy leakage";
    if (k == 2*kMaxAbsor+4) title = "total energy relased : Edep + Eleak";        
    G4int ih = analysisManager->CreateH1(id[k], title, nbins, vmin, vmax);
    analysisManager->SetH1Activation(ih, false);
  }
}
